Mathlib/GroupTheory/FreeGroup/ReducedWords.lean

```diff
@@ -0,0 +1,123 @@
+/-
+Copyright (c) 2025 Bernhard Reinke. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Amir Livne Bar-on, Bernhard Reinke
+-/
+import Mathlib.Data.List.Chain
+import Mathlib.GroupTheory.FreeGroup.Reduce
+
+/-!
+This file defines some extra lemmas for free groups, in particular about (cyclically) reduced words.
+
+## Main declarations
+
+* `FreeGroup.Red.reduced`: the predicate for reduced words
+
+-/
+open List
+
+universe u
+
+variable {α : Type u}
+namespace FreeGroup
+
+variable {L L₁ L₂ : List (α × Bool)}
+
+namespace Red
+
+/-- Predicate asserting that the word `L` admits no reduction steps, i.e., no two neighboring
+elements of the word cancel. -/
+@[to_additive "Predicate asserting the word `L` admits no reduction steps, i.e., no two neighboring
+elements of the word cancel."]
+def reduced (L : List (α × Bool)) : Prop := List.Chain' (fun a b => ¬(a.1 = b.1 ∧ (!a.2) = b.2)) L
+
+@[to_additive (attr := simp)]
+theorem reduced_nil : reduced ([] : List (α × Bool)) := List.chain'_nil
+
+@[to_additive (attr := simp)]
+theorem reduced_singleton {a : (α × Bool)} : reduced [a] := List.chain'_singleton a
+
+@[to_additive (attr := simp)]
+theorem reduced_cons {a b: (α × Bool)} :
+    reduced (a :: b :: L) ↔ ¬(a.1 = b.1 ∧ (!a.2) = b.2) ∧ reduced (b :: L) :=
+  List.chain'_cons
+
+@[to_additive]
+theorem not_step_reduced : reduced L₁ → ¬ Step L₁ L₂
+| red, step => by induction step; simp [reduced] at red
+
+@[to_additive]
+theorem not_step_reduced_iff : reduced L₁ ↔ ∀ L₂, ¬ Step L₁ L₂ where
+  mp h _ := not_step_reduced h
+  mpr hL := by
+    induction L₁ with
+    | nil => exact reduced_nil
+    | cons x l ih =>
+      cases l with
+      | nil => exact reduced_singleton
+      | cons y l' =>
+        rw [reduced_cons]
+        constructor
+        · intro ⟨eq1, eq2⟩
+          obtain ⟨x1, x2⟩ := x
+          obtain ⟨y1, y2⟩ := y
+          simp only at eq1 eq2
+          apply hL l'
+          rw [eq1, ← eq2]
+          apply Step.cons_not
+        · apply ih
+          intro L₂ step
+          apply hL (x :: L₂)
+          exact Step.cons step
+
+@[to_additive]
+theorem reduced_infix : reduced L₂ → L₁ <:+: L₂ → reduced L₁ := Chain'.infix
+
+@[to_additive]
+theorem reduced_min (h : reduced L₁) : Red L₁ L₂ ↔ L₂ = L₁ :=
+  Relation.reflTransGen_iff_eq fun _ => not_step_reduced h
+
+@[to_additive]
+theorem reduced_cons_append_chain {x : α × Bool} {L₁ L₂ : List (α × Bool)} (h : L₁ ≠ []) :
+    reduced (x :: L₁) → reduced (L₁ ++ L₂) → reduced (x :: L₁ ++ L₂) := by
```

Comment:
the first 2 args should be hypotheses, see https://leanprover-community.github.io/contribute/style.html#hypotheses-left-of-colon